File: api/src/layouts/overlay-center.ts
// Layout: Overlay Center - Full background + centered text
import type { Layout, LayoutRenderParams } from '../lib/types';
import { escapeHtml, sanitizeColor, getOverlayGradient, sanitizeUrlForCss, logoPositionStyle, watermarkHtml } from '../lib/html-helpers';

export const overlayCenterLayout: Layout = {
  id: 'overlay-center',
  name: 'Overlay Center',
  categories: ['landscape', 'square'],
  params: [
    { key: 'title', type: 'text', label: 'Title', required: true },
    { key: 'subtitle', type: 'text', label: 'Subtitle' },
    { key: 'bg_color', type: 'color', label: 'Background Color' },
    { key: 'bg_image', type: 'image', label: 'Background Image', searchable: true },
    { key: 'title_color', type: 'color', label: 'Title Color' },
    { key: 'subtitle_color', type: 'color', label: 'Subtitle Color' },
    { key: 'overlay', type: 'select', label: 'Overlay', options: ['none', 'light', 'medium', 'dark'] },
  ],

  render(p: LayoutRenderParams): string {
    const bg = sanitizeColor(p.bg_color, '#1a1a3e');
    const titleColor = sanitizeColor(p.title_color, '#FFD700');
    const subColor = sanitizeColor(p.subtitle_color, '#FFFFFF');
    const title = escapeHtml(p.title);
    const subtitle = escapeHtml(p.subtitle);
    const overlay = getOverlayGradient(p.overlay);

    const safeBgImage = sanitizeUrlForCss(p.bg_image);
    const bgStyle = safeBgImage
      ? `background:${overlay !== 'none' ? overlay + ',' : ''}url('${safeBgImage}') center/cover no-repeat`
      : `background:${bg}`;

    const safeLogo = sanitizeUrlForCss(p.logo);
    const logoHtml = safeLogo
      ? `<img src="${safeLogo}" alt="logo" style="position:absolute;${logoPositionStyle(p.logo_position)};height:8%;object-fit:contain;" crossorigin="anonymous" onerror="this.style.display='none'" />`
      : '';
    const wmHtml = watermarkHtml(p.watermark_url, p.watermark_opacity);

    const subtitleHtml = subtitle
      ? `<p style="color:${subColor};font-weight:500;font-size:clamp(0.9rem,2.5vw,1.8rem);margin-top:0.5em;padding:0 10%;opacity:0.9;font-family:'Be Vietnam Pro',sans-serif;">${subtitle}</p>`
      : '';

    return `<div id="thumbnail" style="width:${p.width}px;height:${p.height}px;${bgStyle};display:flex;flex-direction:column;align-items:center;justify-content:center;text-align:center;position:relative;font-family:'Montserrat',sans-serif;overflow:hidden;">
  ${logoHtml}
  <h1 style="color:${titleColor};font-weight:800;font-size:clamp(1.5rem,5vw,4rem);line-height:1.2;margin:0;padding:0 8%;font-family:'Montserrat',sans-serif;">${title}</h1>
  ${subtitleHtml}
  ${wmHtml}
</div>`;
  },
};

File: server/layouts/overlay_center.go
package layouts

import (
	"fmt"
	"snapkit/engine"
	"snapkit/models"
)

func overlayCenter() *BuiltinLayout {
	return &BuiltinLayout{
		Info: models.LayoutInfo{
			ID:         "overlay-center",
			Name:       "Overlay Center",
			Categories: []string{"landscape", "square"},
			Params: []models.LayoutParam{
				{Key: "title", Type: "text", Label: "Title", Required: true},
				{Key: "subtitle", Type: "text", Label: "Subtitle"},
				{Key: "bg_color", Type: "color", Label: "Background Color"},
				{Key: "bg_image", Type: "image", Label: "Background Image", Searchable: true},
				{Key: "title_color", Type: "color", Label: "Title Color"},
				{Key: "subtitle_color", Type: "color", Label: "Subtitle Color"},
				{Key: "overlay", Type: "select", Label: "Overlay", Options: []string{"none", "light", "medium", "dark"}},
			},
		},
		Render: func(p *engine.RenderParams) string {
			bg := engine.SanitizeColor(p.BgColor, "#1a1a3e")
			titleColor := engine.SanitizeColor(p.TitleColor, "#FFD700")
			subColor := engine.SanitizeColor(p.SubtitleColor, "#FFFFFF")
			title := engine.EscapeHTML(p.Title)
			subtitle := engine.EscapeHTML(p.Subtitle)
			overlay := engine.GetOverlayGradient(p.Overlay)

			safeBg := engine.SanitizeURLForCSS(p.BgImage)
			bgStyle := fmt.Sprintf("background:%s", bg)
			if safeBg != "" {
				if overlay != "none" {
					bgStyle = fmt.Sprintf("background:%s,url('%s') center/cover no-repeat", overlay, safeBg)
				} else {
					bgStyle = fmt.Sprintf("background:url('%s') center/cover no-repeat", safeBg)
				}
			}

			logoHTML := engine.LogoHTML(p.Logo, p.LogoPosition, "8%", "")
			wmHTML := engine.WatermarkHTML(p.WatermarkURL, p.WatermarkOpacity)

			subtitleHTML := ""
			if subtitle != "" {
				subtitleHTML = fmt.Sprintf(`<p style="color:%s;font-weight:500;font-size:clamp(0.9rem,2.5vw,1.8rem);margin-top:0.5em;padding:0 10%%;opacity:0.9;font-family:'Be Vietnam Pro',sans-serif;">%s</p>`, subColor, subtitle)
			}

			return fmt.Sprintf(`<div id="thumbnail" style="width:%dpx;height:%dpx;%s;display:flex;flex-direction:column;align-items:center;justify-content:center;text-align:center;position:relative;font-family:'Montserrat',sans-serif;overflow:hidden;">
  %s
  <h1 style="color:%s;font-weight:800;font-size:clamp(1.5rem,5vw,4rem);line-height:1.2;margin:0;padding:0 8%%;font-family:'Montserrat',sans-serif;">%s</h1>
  %s
  %s
</div>`, p.Width, p.Height, bgStyle, logoHTML, titleColor, title, subtitleHTML, wmHTML)
		},
	}
}
